pkg/api/management/api/alert/notifier_action.go

```diff
@@ -0,0 +1,376 @@
+package alert
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"mime/multipart"
+	"net/http"
+	"net/mail"
+	"net/smtp"
+	"net/textproto"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/rancher/norman/parse"
+	"github.com/rancher/norman/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func NotifierCollectionFormatter(apiContext *types.APIContext, collection *types.GenericCollection) {
+	collection.AddAction(apiContext, "send")
+}
+
+func NotifierFormatter(apiContext *types.APIContext, resource *types.RawResource) {
+	resource.AddAction(apiContext, "send")
+}
+
+func (h *Handler) NotifierActionHandler(actionName string, action *types.Action, apiContext *types.APIContext) error {
+	switch actionName {
+	case "send":
+
+		return h.testNotifier(actionName, action, apiContext)
+	}
+
+	return errors.Errorf("bad action %v", actionName)
+
+}
+
+func (h *Handler) testNotifier(actionName string, action *types.Action, apiContext *types.APIContext) error {
+	actionInput, err := parse.ReadBody(apiContext.Request)
+	if err != nil {
+		return err
+	}
+
+	msg := ""
+	msgInf, exist := actionInput["message"]
+	if exist {
+		msg = msgInf.(string)
+	}
+
+	if apiContext.ID != "" {
+		parts := strings.Split(apiContext.ID, ":")
+		ns := parts[0]
+		id := parts[1]
+		notifier, err := h.Management.Management.Notifiers("").GetNamespaced(ns, id, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		if notifier.Spec.SlackConfig.URL != "" {
+			err := testSlack(notifier.Spec.SlackConfig.URL, msg)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+
+		if notifier.Spec.SMTPConfig.Host != "" {
+			s := notifier.Spec.SMTPConfig
+			return testEmail(s.Host, s.Password, s.Username, int(s.Port), s.TLS, msg, s.DefaultRecipient)
+		}
+
+		if notifier.Spec.PagerdutyConfig.ServiceKey != "" {
+			return testPagerduty(notifier.Spec.PagerdutyConfig.ServiceKey, msg)
+		}
+
+		if notifier.Spec.WebhookConfig.URL != "" {
+			return testWebhook(notifier.Spec.WebhookConfig.URL, msg)
+		}
+
+	} else {
+
+		slackConfigInterface, exist := actionInput["slackConfig"]
+		if exist {
+			slackConfig, ok := slackConfigInterface.(map[string]interface{})
+			if ok {
```

Comment:
if `slackConfig` exists, but !ok, do we want to check for other configs (smptConfig,etc), or do we need to return an error?

Reply:
I think we should return an error